File: backend/app/schemas/mobile.py
from __future__ import annotations

from datetime import date as Date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import (
    AbsenceType,
    AssignmentType,
    SiteStatus,
    ToolIssueReason,
    ToolMaterialCategory,
)


class MobilePerson(BaseModel):
    id: int
    first_name: str | None = None
    last_name: str | None = None
    display_name: str
    phone: str | None = None
    email: str | None = None
    can_sign_measurements_immediately: bool = False


class MobileSite(BaseModel):
    id: int
    site_number: str | None = None
    name: str
    location: str | None = None
    address: str | None = None
    customer: str | None = None
    project_manager: MobilePerson | None = None
    status: SiteStatus
    info: str | None = None
    requires_extra_work_approval: bool = False


class MobileAssignment(BaseModel):
    id: int
    start_date: Date
    end_date: Date
    assignment_type: AssignmentType
    note: str | None = None
    person: MobilePerson
    site: MobileSite


class MobileAssignmentsResponse(BaseModel):
    start_date: Date
    end_date: Date
    assignments: list[MobileAssignment]


class MobileSelfPlanRequest(BaseModel):
    site_id: int
    work_date: Date


class MobilePersonalFileVehicle(BaseModel):
    id: int
    license_plate: str
    manufacturer: str


class MobilePersonalFileHoursAccount(BaseModel):
    current_balance_minutes: int
    last_entry_at: datetime | None = None


class MobileToolIssueSummary(BaseModel):
    id: int
    reason: ToolIssueReason
    status: str
    description: str
    created_at: datetime


class MobilePersonalFileTool(BaseModel):
    id: int
    category: ToolMaterialCategory
    beg_number: str | None = None
    manufacturer: str | None = None
    designation: str
    device_number: str | None = None
    item_date: Date | None = None
    open_issue_reports: list[MobileToolIssueSummary] = Field(default_factory=list)


class MobileToolIssueReportCreate(BaseModel):
    reason: ToolIssueReason
    request_id: UUID


class MobileToolIssueReportRead(BaseModel):
    id: int
    status: str
    created_at: datetime
    message: str
    already_reported: bool = False


class MobilePersonalFileResponse(BaseModel):
    current_year: int
    remaining_vacation_days: int
    total_vacation_days: int
    sick_days: int
    hours_account: MobilePersonalFileHoursAccount
    vehicle: MobilePersonalFileVehicle | None = None
    tool_count: int
    tool_preview: list[MobilePersonalFileTool]


class MobilePersonalFileAbsenceEntry(BaseModel):
    source_id: int
    absence_type: AbsenceType
    start_date: Date
    end_date: Date
    day_count: int = Field(ge=1)


class MobilePersonalFileAbsenceWeek(BaseModel):
    iso_year: int
    iso_week: int
    week_start: Date
    week_end: Date
    entries: list[MobilePersonalFileAbsenceEntry]


class MobilePersonalFileAbsenceResponse(BaseModel):
    year: int
    absence_type: AbsenceType
    remaining_vacation_days: int
    total_vacation_days: int
    taken_vacation_days: int
    sick_days: int
    weeks: list[MobilePersonalFileAbsenceWeek]

File: frontend/src/types/mobile.ts
import type { AssignmentType, SiteStatus } from "./matrix";
import type { ToolMaterialCategory } from "./toolMaterial";

export type MobilePerson = {
  id: number;
  first_name: string | null;
  last_name: string | null;
  display_name: string;
  phone: string | null;
  email: string | null;
  can_sign_measurements_immediately: boolean;
};

export type MobileSite = {
  id: number;
  site_number: string | null;
  name: string;
  location: string | null;
  address: string | null;
  customer: string | null;
  project_manager: MobilePerson | null;
  status: SiteStatus;
  info: string | null;
  requires_extra_work_approval: boolean;
};

export type MobileAssignment = {
  id: number;
  start_date: string;
  end_date: string;
  assignment_type: AssignmentType;
  note: string | null;
  person: MobilePerson;
  site: MobileSite;
};

export type MobileAssignmentsResponse = {
  start_date: string;
  end_date: string;
  assignments: MobileAssignment[];
};

export type MobileAssignmentSiteSummary = {
  site: MobileSite;
  last_assignment_date: string;
};

export type MobileAssignmentSitesResponse = {
  through_date: string;
  sites: MobileAssignmentSiteSummary[];
};

export type MobileAssignmentSiteHistoryResponse = {
  through_date: string;
  site: MobileSite;
  assignments: MobileAssignment[];
};

export type MobilePersonalFileVehicle = {
  id: number;
  license_plate: string;
  manufacturer: string;
};

export type MobilePersonalFileHoursAccount = {
  current_balance_minutes: number;
  last_entry_at: string | null;
};

export type MobilePersonalFileTool = {
  id: number;
  category: ToolMaterialCategory;
  beg_number: string | null;
  manufacturer: string | null;
  designation: string;
  device_number: string | null;
  item_date: string | null;
  open_issue_reports: MobileToolIssueSummary[];
};

export type MobileToolIssueReason = "DEFECTIVE" | "STOLEN";

export type MobileToolIssueSummary = {
  id: number;
  reason: MobileToolIssueReason;
  status: string;
  description: string;
  created_at: string;
};

export type MobileToolIssueReport = {
  id: number;
  status: string;
  created_at: string;
  message: string;
  already_reported: boolean;
};

export type MobilePersonalFile = {
  current_year: number;
  remaining_vacation_days: number;
  total_vacation_days: number;
  sick_days: number;
  hours_account: MobilePersonalFileHoursAccount;
  vehicle: MobilePersonalFileVehicle | null;
  tool_count: number;
  tool_preview: MobilePersonalFileTool[];
};

export type MobilePersonalFileAbsenceType = "vacation" | "sick";

export type MobilePersonalFileAbsenceEntry = {
  source_id: number;
  absence_type: MobilePersonalFileAbsenceType;
  start_date: string;
  end_date: string;
  day_count: number;
};

export type MobilePersonalFileAbsenceWeek = {
  iso_year: number;
  iso_week: number;
  week_start: string;
  week_end: string;
  entries: MobilePersonalFileAbsenceEntry[];
};

export type MobilePersonalFileAbsenceResponse = {
  year: number;
  absence_type: MobilePersonalFileAbsenceType;
  remaining_vacation_days: number;
  total_vacation_days: number;
  taken_vacation_days: number;
  vacation_carryover_days: number;
  sick_days: number;
  weeks: MobilePersonalFileAbsenceWeek[];
};
